pyscf/mcscf/apc.py

```diff
@@ -0,0 +1,410 @@
+# Author: Daniel S. King
+
+'''
+APC Ranked-Orbital Active Space Selection
+If you find this module useful for your work, please consider citing the following:
+
+A Ranked-Orbital Approach to Select Active Spaces for High-Throughput Multireference Computation
+https://doi.org/10.1021/acs.jctc.1c00037
+
+Large-Scale Benchmarking of Multireference Vertical-Excitation Calculations via Automated Active-Space Selection
+https://doi.org/10.1021/acs.jctc.2c00630
+'''
+
+from pyscf.lib import logger
+from pyscf import scf, lib
+import numpy as np
+
+class Chooser():
+    """
+    Chooser Class
+    Implements the ranked-orbital selection scheme outlined in https://doi.org/10.1021/acs.jctc.1c00037
+    Given a set of entropies, will select all orbitals for the active space and then drop the lowest-entropy orbitals
+    until the size constraint max_size is met.
+    
+    Args:
+        orbs: 2D Numpy Array
+            Orbitals to choose from, spanning the entire basis (must be square matrix of coefficients)
+        occ: 1D Numpy Array
+            Orbital occupations for orbs (2,1,0); nactel will be set to the number of electrons in the selected orbitals
+        entropies: 1D Numpy Array 
+            Importance measurment used to rank the orbitals
+        max_size: Int or Tuple
+            Active space size constraint. If tuple, interpreted as (nelecas,ncas), if int interpreted as max # of orbitals
+            
+    Returns:
+        active-space-size, #-active-electrons, orbital-initial-guess, chosen-active-orbital-indeces
+        
+    Example:
+    
+    #Randomly ranked orbitals
+    >>> import numpy as np
+    >>> from pyscf import gto, scf, mcscf
+    >>> from pyscf.mcscf import apc
+    >>> mol = gto.M(atom='H 0 0 0; H 0 0 1', basis='ccpvtz')
+    >>> mf = scf.RHF(mol).run()
+    >>> entropies = np.random.choice(np.arange(len(mf.mo_occ)),len(mf.mo_occ),replace=False)
+    >>> chooser = apc.Chooser(mf.mo_coeff,mf.mo_occ,entropies,max_size=(2,2))
+    >>> ncas, nelecas, casorbs, active_idx = chooser.kernel()
+    >>> mc = mcscf.CASSCF(mf, ncas, nelecas).run(casorbs)
+    """
+    
+    def __init__(self,orbs,occ,entropies,max_size=(8,8),verbose=4):
+        #Check that we have a full set of orbitals:
+        assert(orbs.shape[0] == orbs.shape[1])
+        assert(len(occ) == len(entropies))
+        assert(len(entropies) == orbs.shape[1])
+        
+        self.log = logger.new_logger(lib.StreamObject,verbose)
+        self.orbs = orbs
+        self.occ = np.array(occ)
+        self.entropies = entropies
+        self.max_size = max_size
+        self.verbose = verbose
+        
+    def _ncsf(self,nactel,norbs):
```

Comment:
What's the difference between _ncsf and _calc_ncsf? It's better to have some comments here to explain the difference

Reply:
Have added comments 👍